RA-13-Recipe-Screen

DenisGubernatorov opened the pull request:
1. Реализация экрана рецепта. Доработа лэйата
2. Добавлены лэйауты для элементов ресайклеров
3. Реализация ресайклеров в коде

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Перенос зависимости плагина kotlin-parcelize в toml файл
2) Поправлена ошибка, когда при нажатии кнопки "back", из рецепта, не корректно отображалась шапка фрагмента для списка категорий

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Добавлены лэйауты для вьюх экрана рецепта
2) Мелкий рефакторинг ресурсов

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Добавлен адаптер IngredientsAdapter для recyclerView rvIngredients
2) Изменено получение изображения категории, на основе его url. Используется префикс, для получения большого варианта изображения.
3) При получении изображения используется try_with_resources
4) Небольшие правки в лэйатуе экрана рецепта

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Добавлен адаптер для метода приготовления.
2) Доработка лэйаутов экрана рецепта, ингредиентов и метода приготовления.
3) В ресайклеры добавлен разделитель элементов

DenisGubernatorov commented:
1.  На изображении в шапке рецепта должен быь значок "Избранное" но что-то я в макете не нашел пустого значка такого размера. Есть только маленькие, на кнопку. Использовал его
2. В задаче написано, что nestedscrollview должен  быть  основном контейнером. Но тогда как-то не красиов получается. Шапка тоже скролится. Шапку сделал ститческой. А вот все остальное в nestedscrollview добавил. Или я не очень понял что такое основой контейнер... 
3. Может я как-то не так понимаю. Но у меня ожидаемое взаимное поведение у скроллера и ресайклеров было что экран может скроллится отдельно, а контент ресайклеров отдельно. Ну типа если список ингридентов растягивает свою вьюху на всю видимую область, и, при этом, всеравно полностью не выводится. То оснвной экран долже иметь возможность прокрутиться до "Способ" и сам список ингридиентов тоже должен прокручиваться. Но самое лучшее чего смог добиться это как сечас. Т.е. в "Ингредиенты" и "Способ" выводятся полные списки. А основной экран просто прокручивается. 
4. На моем телефоне отображается с багом.
![image](https://github.com/user-attachments/assets/9cf83701-0191-4415-9d49-d9d963f06e23)
Если название ингридента не помещается в свою вьху по длине и переносится на другую строку, то тогда под вьхой для количества и единиц имзерения появляется основной фон.
P.S.  Походу на полноценный ПР это не тянет :)) Ну хоть начать ошибки обсуждать

aGroupoid reviewed (changes requested):
1. Векторные изображения и отличаются от обычных пнгшек, что они могут принимать любые размеры. Нужно импортировать иконку сердца с нужным размером
2. По заданию подразумевается, что шапка тоже скролится. Можешь оставить как тебе нравится
3. В constrainLayout чтоб всё работало корректно, нужно указывать адекватно по каким объектам ограничена твоя вью. У твоей NestedScrollView нету ограничания по низу,  из-за этого он отрисовывает какие-то свои размеры объекта. Достаточно прописать у NestedScrollView app:layout_constraintBottom_toBottomOf="parent"
4) Сделай в вёрстке чтоб текст с количеством занимал match_parent

Всё что внутри nestedScrollView надо переделывать. Каждый Layout жрёт довольно много памяти, в твоём случае можно обойтись только рутовым constrainLayout и одним linearLayout внутри скроллВью. Сейчас всё это очень не оптимизировано работает. 
Также глянь как всё это выглядит в фигме, сейчас не хватает отступов у по вертикали и горизонтали. Между текстом и ресайклер вью отступы не те

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -21,19 +27,59 @@ class RecipeFragment : Fragment() {
+            MaterialDividerItemDecoration(
+                requireContext(),
+                DividerItemDecoration.VERTICAL
+            )
```
Сейчас divider не соответсвует дизайну, настрой эти параметры для него
isLastItemDecorated
dividerInsetStart 
dividerInsetEnd 
dividerColor
dividerThickness

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -21,19 +27,59 @@ class RecipeFragment : Fragment() {
+        val methodAdapter = MethodAdapter(method, requireContext())
+        binding.rvMethod.adapter = methodAdapter
+        binding.rvMethod.addItemDecoration(
+            MaterialDividerItemDecoration(
```
Зачем создавать два одинаковых объекта для дивайдеров, если можно использовать один в двух местах

aGroupoid commented on app/src/main/res/values/strings.xml:
```diff
@@ -9,5 +9,12 @@
     <string name="get_recipes_error">Ошибка. Мы уже работаем над этим</string>
+    <string name="recipe_icon_favorite">Значок избранное</string>
+    <string name="ingredients_title">Ингредиенты</string>
+    <string name="method_title">Способ</string>
```
В макете другая надпись

aGroupoid commented on app/src/main/res/layout/recipe_fragment.xml:
```diff
@@ -1,14 +1,123 @@
+        android:layout_marginStart="@dimen/margin_full_16_dp"
+        android:layout_marginBottom="@dimen/margin_full_16_dp"
+        android:background="@drawable/header_title_shape"
+        android:padding="@dimen/cat_header_padding"
```
Какой-то оромный список параметров здесь, многое можно вынести в стиль. Также лучше поменять название стиля, он не только в categoy используется. Всё писать в одном стиле snake_case или camelCase

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Добавлен корректный значок "Избранное" для шапки рецепта. Поправлен лэйаут
2) Переименование стиля основного текста в шапке

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Сделан отдельный стиль для названий в шапках. Фрагменты переведены на него
2) Сделан отдельный стиль для разделов во фрагменте для рецепта. Добавлен для соответствующих вьюх

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Настройка MaterialDividerItemDecoration

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Перекомпоновка NestedScrollView

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Экран рецепта, добавлен отступ от кнопок

aGroupoid commented on app/src/main/res/layout/recipe_fragment.xml:
```diff
@@ -40,6 +40,7 @@
     <androidx.core.widget.NestedScrollView
         android:layout_width="match_parent"
         android:layout_height="@dimen/_0dp"
+        android:layout_marginBottom="@dimen/recipe_screen_bottom_13dp"
```
Не понимаю смысла этого Марджина,  пустота между кнопками и фрагментами только выросла

aGroupoid reviewed (changes requested):
Экран должен выглядит 1 в 1 как на макете их фигмы. Как минимум в item нужно добавить нужные padding, у заднего фона recyclerView должны быть скосы по углам, как ты это раньше делал для текста в шапке

DenisGubernatorov committed:
RA-13-Recipe-Screen

1) Правка паддингов в ресайклерах
2) Добавлено скругление углов ресайклеров

aGroupoid commented on app/src/main/res/layout/recipe_fragment.xml:
```diff
@@ -1,14 +1,111 @@
+            android:orientation="vertical">
+
+            <LinearLayout
+                android:id="@+id/ingredients_block_header"
```
Не вижу смысла в этом linearLayout, у тебя уже есть textView "Список рецептов" , которая нормально отображается без linearLayout. Каждый лишний лейаут -  дополнительная нагрузка на систему

aGroupoid reviewed (approved):
Молодец, всё верно!
